[DEVC-166] OTA daemon using libnetwork

pmiettinen opened the pull request:
https://swift-nav.atlassian.net/wiki/spaces/DEVC/pages/42663976/Duro+OTA+-+Design+Doc

benjaminaltieri commented:
@mfine any insights?

benjaminaltieri commented:
@pmiettinen for now try to use the mock server in [swift-nav/piksi-releases](https://github.com/swift-nav/piksi-releases/blob/master/upgrade/upgrade-server.hs) and the command mentioned [here](https://swift-nav.atlassian.net/browse/DEVC-167?focusedCommentId=15083&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15083)

mfine commented:
@pmiettinen wow, nice approach - great that the FD can be re-used as an interface into libnetwork where the file is the download location vs. a pipe. Nice job!

Yeah, the upgrader only has 2 devices currently, and does not do default advertising:

https://github.com/swift-nav/skylark/blob/master/upgrader/conf/production.yaml

```yaml
upgrader-images:
  -
    device-uid: 21939d22-15e0-4d77-a2b9-fad862aeb962
    version:    v1.4.10
    sha256:     4dadd520ea69b6a01f185ddb1ae07b3643eda735cf290d38e7b27891092d8d49
    bucket:     swiftnav-releases
    key:        piksi_buildroot/v1.4.10/v3/prod/PiksiMulti-v1.4.10.bin
  -
    device-uid: 9bca6a07-41ea-4cf3-a992-e398a8daebe2
    version:    v1.4.10
    sha256:     4dadd520ea69b6a01f185ddb1ae07b3643eda735cf290d38e7b27891092d8d49
    bucket:     swiftnav-releases
    key:        piksi_buildroot/v1.4.10/v3/prod/PiksiMulti-v1.4.10.bin
```

I can your device listed in this PR. It's kind of a sucky way of doing things for now. You can also use the referenced dummy server to mock it out.

Nice job! Exciting to see this coming in

pmiettinen commented:
Thanks for the comments. Ultimately it was the uuid not matching was server was offering. I tested with `21939d22-15e0-4d77-a2b9-fad862aeb962` and got a valid response.

The uuid in the original post is just some random I took from the OTA document for testing. The Piksi I actually use in my bench testing has uuid `A4956DEF-A56A-4495-97C2-CFF15403C329`. @mfine is it possible to add this to the actual server for example next week so I can test as close to reality as possible?

I got the mock up server running easily but it's giving error `MissingFileError "/mnt/users/pasi/.aws/credentials"` when receiving a request. I don't have valid AWS credentials at my disposal.

Anyway, I can hack something together with these pieces for today to get forward with the ota_daemon debugging.

mfine commented:
I can get your device into the production service - let me know if there are other devices you'd like to get in. The mock server will actually sign credentials around a stored image in S3, and then you can download the signed images. That exercises the s3 signing and downloading path. But I can make another simpler server that just lets you pass whatever url you want and then you can pass that in and download from that url.

pmiettinen commented:
Yep I was thinking along the same lines. First time ever touching Haskell code, ended up with this:

```
#!/usr/bin/env stack
{- stack
    runghc
    --package aeson
    --package http-types
    --package text
    --package time
    --package turtle
    --package wai
    --package warp
 -}

{-# LANGUAGE OverloadedStrings #-}

import Control.Monad.Trans.AWS
import Data.Aeson
import Data.Text.Encoding
import Data.Time
import Network.HTTP.Types
import Network.Wai
import Network.Wai.Handler.Warp
import Turtle                   (Parser, Text, optInt, optText, options)

application :: Text -> Text -> Text -> Text -> Text -> Application
application v s b k u _req resp = do
  resp $ responseLBS status200 [ (hContentType, "application/json") ] $ encode $ object
    [ "version" .= v
    , "sha256"  .= s
    , "url"     .= u
    ]

parser :: Parser (Port, Text, Text, Text, Text, Text)
parser = (,,,,,) <$> p <*> v <*> s <*> b <*> k  <*> u
  where
    p = optInt  "port"    'p' "Port"
    v = optText "version" 'v' "Version"
    s = optText "sha256"  's' "SHA-256"
    b = optText "bucket"  'b' "Bucket"
    k = optText "key"     'k' "Key"
    u = optText "url"     'u' "URL"

main :: IO ()
main = do
  (p, v, s, b, k, u) <- options "Upgrade Server" parser
  run p $ application v s b k u
```

mfine commented:
Yeah, nice - exactly - you can throw away the bucket and key as parameters since you're not using them

mfine commented:
Here's a no-sign version:

https://github.com/swift-nav/piksi-releases/blob/master/upgrade/upgrade-server-no-sign.hs

```haskell
#!/usr/bin/env stack
{- stack
    runghc
    --package aeson
    --package http-types
    --package text
    --package time
    --package turtle
    --package wai
    --package warp
 -}

{-# LANGUAGE OverloadedStrings #-}

import Data.Aeson
import Data.Text.Encoding
import Data.Time
import Network.HTTP.Types
import Network.Wai
import Network.Wai.Handler.Warp
import Turtle                   (Parser, Text, optInt, optText, options)

application :: Text -> Text -> Text -> Application
application v s u _req resp = do
  resp $ responseLBS status200 [ (hContentType, "application/json") ] $ encode $ object
    [ "version" .= v
    , "sha256"  .= s
    , "url"     .= u
    ]

parser :: Parser (Port, Text, Text, Text)
parser = (,,,) <$> p <*> v <*> s <*> u
  where
    p = optInt  "port"    'p' "Port"
    v = optText "version" 'v' "Version"
    s = optText "sha256"  's' "SHA-256"
    u = optText "url"     'u' "URL"

main :: IO ()
main = do
  (p, v, s, u) <- options "Upgrade Server" parser
  run p $ application v s u
```

pmiettinen commented:
Mockup server approach is working fine. Learning the libnetwork and libcurl usage and the pieces are falling in place. JSON parsing to be done still. No active blockers at the moment.

pmiettinen commented:
Thanks for the guidance @silverjam, I'll prototype the JSON part to get the end-to-end operation running and then start working on the tweaks.

pmiettinen commented:
Changed to `runit_service` based approach. First successful end-to-end OTA update done against mock-up server including running `upgrade_tool` and `do_sbp_msg_reset`.

Quality and robustness work still to be done.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018091419-13-g07d93e3a
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=681947)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-13-g07d93e3a&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-13-g07d93e3a&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018091419-13-g07d93e3a
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018091419-13-g07d93e3a)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018091419-13-g07d93e3a/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018091419-13-g07d93e3a

silverjam commented:
👍

pmiettinen commented:
Added `run_with_stdin_file` functionality as requested. Added robustness to version number checking. Some leftover stuff for tomorrow left, but I'd expect tomorrow I can remove the WIP tag.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018091419-14-g6a879149
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=683348)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-14-g6a879149&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-14-g6a879149&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018091419-14-g6a879149
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018091419-14-g6a879149)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018091419-14-g6a879149/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018091419-14-g6a879149

pmiettinen commented:
WIP status removed. Ready for review comments.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018091419-16-g1dee116f
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=684889)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-16-g1dee116f&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-16-g1dee116f&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018091419-16-g1dee116f
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018091419-16-g1dee116f)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018091419-16-g1dee116f/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018091419-16-g1dee116f

benjaminaltieri commented:
Sorry to blow up your PR here, but if you rebase against `master` and run `make clang-format` you should be able to resolve any conflicts introduced by #834

silverjam commented on package/libpiksi/libpiksi/src/util.c:
```diff
@@ -372,3 +383,122 @@ void setup_sigterm_handler(void (*handler)(int signum, siginfo_t *info, void *uc
+  return !(-1 == ret && errno == ESPIPE);
+}
+
+int run_with_stdin_file(const char *input_file,
```
Let's add a few simple unit tests for this helper:
- Test overflow conditions
- Test no input file
- Test with an input file

pmiettinen replied:
Any convenient example on the unit test framework or what did you have in mind? Need test some functions also...

silverjam replied:
Seems you already found some unit testing examples, these projects also have unit tests:
- https://github.com/swift-nav/piksi_buildroot/tree/master/package/ports_daemon/ports_daemon/test
- https://github.com/swift-nav/piksi_buildroot/tree/master/package/sbp_fileio_daemon/sbp_fileio_daemon/test
- https://github.com/swift-nav/piksi_buildroot/tree/master/package/sbp_settings_daemon/sbp_settings_daemon/test

This commit is a good example of the transformation that typically happens when adding unit tests to a buildroot package:
- https://github.com/swift-nav/piksi_buildroot/commit/a5883f8dfa155468376c809b6da58391dcd35bea

silverjam commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+  int ret = ota_cmp_version(&offered_parsed, &current_parsed);
+
+  if (ret > 0) {
+    piksi_log(LOG_INFO, "OTA new version available: %s", offered);
```
`"New version available via OTA: %s"`

pmiettinen replied:
Done.

pmiettinen replied:
Done.

silverjam commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+  struct json_object *fjson = json_tokener_parse(fmap);
+
+  struct json_object *json_url = json_object_object_get(fjson, "url");
+  strncpy(parsed_resp->url,
```
Switch to snprintf and log a warning if we overflow these buffers?

pmiettinen replied:
Replaced `CHECKED_SPRINTF` macro with `snprintf_assert` and `snprintf_warn` wrapper functions.

Not sure if I followed to the fullest with the latter part of the comment so I didn't act on in. Please elaborate if further actions are desired.

silverjam replied:
We can leave this out of this PR if you don't feel it's appropriate but I was thinking it might be a good idea to have some defensive code in the download of resources from the remote system.  Particularly in `network_download_write` we can have 2 new parameters has part of the `ctx` struct a counter of the current amount of data downloaded and a limit.  For JSON blob downloads, this should be limited to 4k, for firmware download it should limited to 30mb.

silverjam replied:
The fixes to `CHECKED_SPRINTF` look good though, thanks.

pmiettinen replied:
I think I'll leave that to followup-PR. There's already quite a few diversions in this PR already and I'd like to get this wrapped inside this sprint if possible.

pmiettinen replied:
Created an issue https://swift-nav.atlassian.net/projects/DEVC/issues/DEVC-1141 and start looking into this.

silverjam commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+  return ret;
+}
+
+static int ota_parse_response(ota_resp_t *parsed_resp) {
```
Return bool

pmiettinen replied:
Done

silverjam commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+  return 0;
+}
+
+static int ota_create_files(int *fd_resp, int *fd_img)
```
Return bool

pmiettinen replied:
Done

silverjam commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+  }
+}
+
+static int ota_client_loop(void)
```
Return bool

pmiettinen replied:
Done

silverjam commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+  int fd_resp = -1;
+  int fd_img = -1;
+
+  while (!libnetwork_shutdown_signaled(nw_ctx) && do_sigtimedwait(&params)) {
```
Will `do_sigtimedwait` return when SIGTERM/SIGINT are received?  Otherwise it seems like we won't terminate until `do_sigtimedwait` fires and we have a chance to check `libnetwork_shutdown_signaled`.

pmiettinen replied:
From https://linux.die.net/man/2/sigtimedwait
```
EINTR

The wait was interrupted by a signal handler; see signal(7). (This handler was for a signal other than one of those in set.)
```

In order for the loop to continue `sigtimedwait` needs to exit with `EAGAIN`.

Also bench testing indicates that behavior is as we expect or want it to be.

silverjam replied:
👍 thanks

pmiettinen commented on package/libnetwork/src/libnetwork.c:
```diff
@@ -897,32 +917,43 @@ static void network_request(network_context_t* ctx, CURL *curl)
   curl_easy_setopt(curl, CURLOPT_BUFFERSIZE,        RECV_BUFFER_SIZE);
 
   while (true) {
-
```
Done

pmiettinen commented on package/libnetwork/src/libnetwork.c:
```diff
@@ -1093,3 +1124,52 @@ void skylark_upload(network_context_t* ctx)
+void ota_enquire(network_context_t *ctx)
+{
+  CURL *curl = network_setup(ctx);
+  if (curl == NULL) {
```
Seems like `network_setup` already logs errors for all the cases where `NULL` is returned. I made the error messages a bit more verbose.

silverjam replied:
👍

pmiettinen commented on package/libnetwork/src/libnetwork.c:
```diff
@@ -1093,3 +1124,52 @@ void skylark_upload(network_context_t* ctx)
+  device_fw_version_get(fw_buf, sizeof(fw_buf));
+  count =
+      snprintf(fw_hdr_buf, sizeof(fw_hdr_buf), "Current-Version: %s", fw_buf);
+  assert((size_t)count < sizeof(uuid_hdr_buf));
```
Classic copy-pasta mistake fixed.

pmiettinen commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+    return 1;
+  }
+
+  if ((status = libnetwork_set_debug(ctx, false)) != NETWORK_STATUS_SUCCESS) {
```
Done.

pmiettinen commented on package/libpiksi/libpiksi/include/libpiksi/util.h:
```diff
@@ -131,8 +146,70 @@ typedef void (*child_exit_fn_t)(pid_t pid);
+ *
+ * @return  0 if signal was received, 1 if timeout was reached
+ */
+int do_sigtimedwait(sigwait_params_t *params);
```
Fair enough.

pmiettinen commented on package/libpiksi/libpiksi/include/libpiksi/util.h:
```diff
@@ -131,8 +146,70 @@ typedef void (*child_exit_fn_t)(pid_t pid);
+
+/**
+ * @brief   Check if file descriptor is file
+ * @details lseek will fail with ESPIPE fd is associated with a pipe, socket or
```
Definitely an improvement.

pmiettinen commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+#define OTA_TIMEOUT_JITTER \
+    ((rand() % (OTA_TIMEOUT_JITTER_MAX + 1 - OTA_TIMEOUT_JITTER_MIN) + \
+    OTA_TIMEOUT_JITTER_MIN) / 100.f)
+#define OTA_TIMEOUT_S round(OTA_TIMEOUT_AVG_S * OTA_TIMEOUT_JITTER)
```
Done.

pmiettinen commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+
+  if (ret) {
+    piksi_log(LOG_ERR, "SHA256 mismatch");
+    piksi_log(LOG_ERR, "%s", sha);
```
Done.

pmiettinen commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+    }
+  }
+
+  char *digits = strdup(str + idx);
```
`strtok_r` modifies the buffer.

pmiettinen commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,529 @@
+/* TODO: check that token includes only digits */
+#define CHECK_TOKEN(version) do {                                   \
+  if (token) {                                                      \
+    version = atoi(token);                                          \
```
`strtoul` doc says that it returns zero for non-valid conversion and since `"0"` is a valid version string I think it won't work for us. I added helper `str_digits_only`.

pmiettinen commented:
@silverjam thanks for the detailed review, I got through maybe ~half of them, will continue on Monday.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018091419-32-g07435d57
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=686719)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-32-g07435d57&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-32-g07435d57&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018091419-32-g07435d57
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018091419-32-g07435d57)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018091419-32-g07435d57/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018091419-32-g07435d57

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018091419-32-g07435d57
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=686773)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-32-g07435d57&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018091419-32-g07435d57&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018091419-32-g07435d57
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018091419-32-g07435d57)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018091419-32-g07435d57/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018091419-32-g07435d57

pmiettinen commented on package/libpiksi/libpiksi/src/util.c:
```diff
@@ -48,7 +49,10 @@
+#define PIPE_READ_SIDE 0
+#define PIPE_WRITE_SIDE 1
+
+int file_read_string(const char *filename, char *str, size_t str_size)
```
Applied based on https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=87152445

silverjam replied:
👍 unit test here would be great

pmiettinen commented:
@silverjam I have now addressed all comments except the one concerning unit tests. Which is also a bigger one as I definitely would like to add unit tests for multiple functions. But I am not sure how to proceed. My investigations indicate that we don't currently have a functioning unit test framework in `piksi_buildroot`, is this correct? If correct, where to start? If not correct, any example how to start adding unit tests?

I am currently building `host-image` and will continue my investigations there.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-17-gfa43d834
There was an error using the HITL API to kick off smoke tests for this commit. See https://travis-ci.org/swift-nav/piksi_buildroot/builds/432381200.

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-17-gfa43d834
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-17-gfa43d834)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-17-gfa43d834/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-17-gfa43d834

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-18-g891bf85a
There was an error using the HITL API to kick off smoke tests for this commit. See https://travis-ci.org/swift-nav/piksi_buildroot/builds/432382753.

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-18-g891bf85a
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-18-g891bf85a)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-18-g891bf85a/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-18-g891bf85a

pmiettinen commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -0,0 +1,535 @@
+  const char *cmd = "sha256sum";
+  char *const argv[3] = {"sha256sum", OTA_IMAGE_FILE_PATH, NULL};
+  char sha[65] = {0};
+  int ret = run_with_stdin_file(NULL, cmd, argv, sha, sizeof(sha));
```
Just noticed while testing that `sha256sum` is left in zombie state after exit. It's easily missed as the upgrade does reboot. Indicates that child exit code is not handled correctly.

pmiettinen replied:
Fixed by adding `wait(NULL)` for parent.

silverjam replied:
👍

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-19-g1eeec881
There was an error using the HITL API to kick off smoke tests for this commit. See https://travis-ci.org/swift-nav/piksi_buildroot/builds/432400195.

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-19-g1eeec881
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-19-g1eeec881)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-19-g1eeec881/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-19-g1eeec881

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-21-gcffd8c3c
There was an error using the HITL API to kick off smoke tests for this commit. See https://travis-ci.org/swift-nav/piksi_buildroot/builds/432422086.

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-21-gcffd8c3c
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-21-gcffd8c3c)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-21-gcffd8c3c/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-21-gcffd8c3c

pmiettinen commented:
Found the `run_libpiksi_tests`.. Having trouble building the host image because of our `NFS` server which squashes root user (chroot etc..)

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-22-g3ae03423
There was an error using the HITL API to kick off smoke tests for this commit. See https://travis-ci.org/swift-nav/piksi_buildroot/builds/432428379.

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-22-g3ae03423
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-22-g3ae03423)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-22-g3ae03423/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-22-g3ae03423

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-22-g3ae03423
There was an error using the HITL API to kick off smoke tests for this commit. See https://travis-ci.org/swift-nav/piksi_buildroot/builds/432442208.

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-22-g3ae03423
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-22-g3ae03423)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-22-g3ae03423/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-22-g3ae03423

pmiettinen commented:
After couple of hours crash course to chroot world I was able to run libpiksi tests with

`proot -r ~/swiftnav/piksi_buildroot/buildroot/host_output/target -b ~/swiftnav/piksi_buildroot/buildroot/host_output/build/libpiksi-0.1/test/ run_libpiksi_tests`

I think I can work with this to get the unit tests running.

I'll try to work with our IT support regarding the NFS problems. It is preventing the docker builds also.

silverjam commented:
@pmiettinen if the unit tests don't require the chroot, we can drop it, I've used other patterns to achieve mostly the same end result:

- https://github.com/swift-nav/piksi_buildroot/blob/silverjam/raw-uds-nox/package/endpoint_router/endpoint_router.mk#L27
- https://github.com/swift-nav/piksi_buildroot/blob/silverjam/raw-uds-nox/package/endpoint_router/endpoint_router/test/endpoint_router_test.cc#L268

silverjam commented on package/ota_daemon/src/ota_daemon.c:
```diff
@@ -43,10 +43,13 @@
-   / 100.f)
-#define OTA_TIMEOUT_S round(OTA_TIMEOUT_AVG_S *OTA_TIMEOUT_JITTER)
+
+static inline int ota_timout_s(void)
```
Spelling nit: `ota_timout_s` -> `ota_timeout_s`

pmiettinen replied:
Typo fixed.

silverjam commented on package/libnetwork/src/libnetwork.c:
```diff
@@ -49,7 +49,7 @@
 
 /** How large to configure the recv buffer to avoid excessive buffering. */
-const long RECV_BUFFER_SIZE = 4096L;
+const long RECV_BUFFER_SIZE = 16 * 1024L;
```
We can leave it if it's faster and doesn't break ntrip/skylark (I don't suspect that it would though).

pmiettinen commented:
Filesystem problems should be cleared now (currently running docker and then host-image builds) and tomorrow I can actually start working on the unit tests.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-31-g2a0dc5fe
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=694348)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092118-31-g2a0dc5fe&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092118-31-g2a0dc5fe&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-31-g2a0dc5fe
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-31-g2a0dc5fe)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-31-g2a0dc5fe/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-31-g2a0dc5fe

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-31-gcf38f9ba
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=694415)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092118-31-gcf38f9ba&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092118-31-gcf38f9ba&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092118-31-gcf38f9ba
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092118-31-gcf38f9ba)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092118-31-gcf38f9ba/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092118-31-gcf38f9ba

pmiettinen commented:
First set of unit tests. More incoming tomorrow.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092118-32-g3d7d2db2
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=694617)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092118-32-g3d7d2db2&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092118-32-g3d7d2db2&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092623-28-g9cc30582
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=696144)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-28-g9cc30582&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-28-g9cc30582&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092623-28-g9cc30582
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092623-28-g9cc30582)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092623-28-g9cc30582/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092623-28-g9cc30582

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092623-29-g31e69d30
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=696196)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-29-g31e69d30&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-29-g31e69d30&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092623-29-g31e69d30
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092623-29-g31e69d30)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092623-29-g31e69d30/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092623-29-g31e69d30

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092623-29-g31e69d30
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=696237)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-29-g31e69d30&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-29-g31e69d30&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092623-29-g31e69d30
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092623-29-g31e69d30)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092623-29-g31e69d30/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092623-29-g31e69d30

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092623-30-g507418e7
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=696347)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-30-g507418e7&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-30-g507418e7&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092623-30-g507418e7
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092623-30-g507418e7)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092623-30-g507418e7/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092623-30-g507418e7

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092623-30-g507418e7
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=696380)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-30-g507418e7&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092623-30-g507418e7&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092623-30-g507418e7
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092623-30-g507418e7)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092623-30-g507418e7/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092623-30-g507418e7

silverjam commented:
@benjaminaltieri any comments before we merge?

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018092718-30-g198417d3
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=697368)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092718-30-g198417d3&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018092718-30-g198417d3&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018092718-30-g198417d3
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018092718-30-g198417d3)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018092718-30-g198417d3/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018092718-30-g198417d3

pmiettinen commented:
https://travis-ci.org/swift-nav/piksi_buildroot/builds/434463620
`release`, `nano` and `sdk` builds failed due to network problems but they were fine in the previous build here https://travis-ci.org/swift-nav/piksi_buildroot/builds/434096842 so going to presume it's fine after the last unit test fixes.
